### PR TITLE
feat(snownet): automatically discover host candidates

### PR DESCRIPTION
**Pull request by thomaseizinger**
I had an idea that is quite a big deal. Instead of manually discovering host candidates by iterating interfaces and adding all of them, we automatically generate host candidates every time we receive traffic on from a certain interface. Initially, you might think that this is a catch-22: How do we generate traffic without having host candidates? The answer is: relays!

When we initiate a new connection using `snownet`, we add a list of STUN and TURN servers. We will immediately attempt to talk to both of them, sending STUN bindings to the former and ALLOCATE requests to the latter. As the replies come in, we know, which interface they have been received on. That particular interface is an excellent `host` candidate because we've already proven connectivity to a STUN or TURN server. In fact, for hole-punching, we will need to send traffic via that same interface to reach the gateway (otherwise the `srflx` candidate won't work anyway).

There is only one "edge"-case in which this doesn't work: When you want to make a connection between a client and a gateway on the same subnet yet without connectivity to a relay. At that point, I'd argue that your network topology is broken anyway. If you can't talk to a relay, you probably also cannot talk to the portal, meaning the signaling protocol also doesn't work.

### CODE REVIEW

**Comment by jamilbk** [comment]
> There is only one "edge"-case in which this doesn't work: When you want to make a connection between a client and a gateway on the same subnet yet without connectivity to a relay. At that point, I'd argue that your network topology is broken anyway. If you can't talk to a relay, you probably also cannot talk to the portal, meaning the signaling protocol also doesn't work.

This is a very common edge case. It will happen any time a remote worker comes into the office for example. It also happens when folks are evaluating whether to use Firezone or not -- one typical test would be to install a gateway on another machine and then ping it. It will also happen for servers with headless clients on them talk to other servers (resources) in the same VPC.

Would that mean all of those connections need to go through a Relay? That would be... not ideal.

I know I'm probably not fully grasping why we can't gather host address directly from interfaces, but I'll give the discussion another read.

But unless we run many more Relays I don't think this will be a competitive enduser experience (or I'm not fully understanding)

**Review by jamilbk — CHANGES_REQUESTED** [review]
If my concerns in [this comment](https://github.com/firezone/firezone/pull/3528#issuecomment-1924002060) are valid, then I want to hold off merging this until we work through the edge cases more.

The edge cases I mentioned are not theoreticals -- we have customers today with these use cases and have asked how we handle them (let me know if it would be helpful to explain further on call)

File: rust/connlib/snownet/src/node.rs [hunk 1]
```diff
@@ -766,21 +793,8 @@ where
         allowed_stun_servers: &HashSet<SocketAddr>,
         allowed_turn_servers: &HashSet<SocketAddr>,
     ) {
-        for local in self.local_interfaces.iter().copied() {
```
Comment (jamilbk) [inline rust/connlib/snownet/src/node.rs:769]: I'm a bit confused as to why this wouldn't have gathered the host candidate used to connect to the relay as well. 

Am I understanding the problem correctly?:

- local interfaces are gathered
- A new socket is opened to the Relay (using source address `local` which is `0.0.0.0`?)
- We gather the resulting socket addrs (`0.0.0.0:8452`)

These are all useless because we can't see the actual source socketaddr the OS picked for us?

**Comment by jamilbk** [comment]
Ah, I think I'm understanding a bit better now (coffee helps!).

This PR would essentially ensure that any data plane communication only happens on the interface used to reach the Relay, an external service.

This is probably OK for the "workers in office" use case and possible "evaluating in my homelab", but not great for server-server communication. Those are typically "multi-homed" to have an interface for default egress and then another attached to the VPC. I first thought maybe this isn't an issue because hairpinning could still work off their shared NAT gateway, but I'm not so sure AWS or other cloud providers and data centers will support that. That means these servers will "talk to each other" using their public IP addresses, which won't be ideal for latency or bandwidth costs.

So, back to the original problem then. I'll chat with @conectado today about it and see if something comes up.

I think being able to use host candidates that are not the default egress interface will soon be a requirement, if it's not already. We can kick the can down the road, but we have full control now right? ;-). We can make this an excellent (read: competitive) experience for customers.

**Comment by conectado** [comment]
> Ah, I think I'm understanding a bit better now (coffee helps!).
> 
> This PR would essentially ensure that any data plane communication only happens on the interface used to reach the Relay, an external service.
> 
> This is probably OK for the "workers in office" use case and possible "evaluating in my homelab", but not great for server-server communication. Those are typically "multi-homed" to have an interface for default egress and then another attached to the VPC. I first thought maybe this isn't an issue because hairpinning could still work off their shared NAT gateway, but I'm not so sure AWS or other cloud providers and data centers will support that. That means these servers will "talk to each other" using their public IP addresses, which won't be ideal for latency or bandwidth costs.
> 
> So, back to the original problem then. I'll chat with @conectado today about it and see if something comes up.
> 
> I think being able to use host candidates that are not the default egress interface will soon be a requirement, if it's not already. We can kick the can down the road, but we have full control now right? ;-). We can make this an excellent (read: competitive) experience for customers.

Adding to this, there's another problem with this approach, we can't discover the address of any interface that isn't used. Meaning, it's not only interfaces that can't reach the relays that aren't used as local candidates, it's also interfaces that could reach the relays but are not prioritized by the routing table.

@thomaseizinger I think it's better to leave things as they are right now, we're losing a lot of conditions where we should be able to hole-punch for a bit of simplicity on the API

File: rust/connlib/snownet/src/node.rs (hunk 1 above)
Comment (thomaseizinger) [reply rust/connlib/snownet/src/node.rs:769]: Yep, `0.0.0.0` doesn't pick an address for you, it means all interfaces.

**Comment by thomaseizinger** [comment]
> @thomaseizinger I think it's better to leave things as they are right now, we're losing a lot of conditions where we should be able to hole-punch for a bit of simplicity on the API

I'd like to challenge the "a lot of conditions" bit. Can you draw a network topology where we don't have connectivity with this apptoach?

I'd like to suggest that we continue down this path. The behaviour suggested here is still correct. Traffic coming in on a socket 100% gives us a viable host candidate. We can always add another API to manually add them later. But before we do that, I'd like to understand the network topology where this doesn't work and be able to write a test for it instead of blindly grabbing candidates from all interfaces.

If you want, we can also put the API back and just not use it for now but I'd like to merge this PR because it is a step forward.

**Comment by thomaseizinger** [comment]
> but not great for server-server communication. Those are typically "multi-homed" to have an interface for default egress and then another attached to the VPC. I first thought maybe this isn't an issue because hairpinning could still work off their shared NAT gateway, but I'm not so sure AWS or other cloud providers and data centers will support that.

Sone questions / points:

- Server-server in the same VPC means "server-headless client" right? Two gateways never connect to each other, right? I didn't know that we want to support this already (and that it is a usecase at all running them in the same VPC).
- Only party needs to add an extra host candidate for this to work. For example, it is probably better to add functionality to the gateway to iterate all its interfaces and add additional candidates but not do this on the client where we might have stuff like docker networks, other VPNs etc.

Do we agree that this PR is an improvement and that we can always add an API back that manually adds host candidates?

**Comment by conectado** [comment]
> I'd like to challenge the "a lot of conditions" bit. Can you draw a network topology where we don't have connectivity with this apptoach?

We will always have connectivity due to relayed connections but we do lose direct connections, in these 2 cases:
* 2 hosts in a local network through a given iface, yet they connect to WAN through a separate interface that due to symmetric nat can't connect to each other
* The LAN interfaces can even be connected to WAN but just no the default route to reach the relays

> Do we agree that this PR is an improvement and that we can always add an API back that manually adds host candidates?

I'm not clear on what we gain, save for having to walk through local interfaces which we know that we can do

**Comment by thomaseizinger** [comment]
> * 2 hosts in a local network through a given iface, yet they connect to WAN through a separate interface that due to symmetric nat can't connect to each other

They also must each go through _separate_ WAN interfaces to the relay, otherwise the router would directly connect them.

You are right that in this case, we don't discover a more optimal connection. Some notes though:

- That is not hole-punching.
- Is that really a common enough topology to say we are missing "a lot of conditions"?
- Like I said in the previous comment, adding all interfaces on the gateway is enough to fix it. We can avoid a lot of (likely rubbish) candidates on the client.

> > Do we agree that this PR is an improvement and that we can always add an API back that manually adds host candidates?
> 
> I'm not clear on what we gain, save for having to walk through local interfaces which we know that we can do

- Error handling within `snownet` is improved. The error condition of "you gave me data on a socket I don't know about" simply no longer exists. (This reminds me I need to remove the error variant).
- No combinatoric explosion of candidate pairs.

I am happy to add the API back. Then this PR is a strict improvement over what we had before. I'll also try to model the above network topology as a docker-compose test.

**Comment by thomaseizinger** [comment]
> * 2 hosts in a local network through a given iface

This might be a silly question but what is the point of using firezone in this topology? You are already in the same network as your gateway so you should be able to reach all the resources.

**Comment by conectado** [comment]
> They also must each go through _separate_ WAN interfaces to the relay, otherwise the router would directly connect them.

Only if hair-pinning is supported

> You are right that in this case, we don't discover a more optimal connection. Some notes though:
> 
>     * That is not hole-punching.
> 
>     * Is that really a common enough topology to say we are missing "a lot of conditions"?
> 
>     * Like I said in the previous comment, adding all interfaces on the gateway is enough to fix it. We can avoid a lot of (likely rubbish) candidates on the client.

but we need to use local candidates from both to achieve this, or am I missing something? 

> 
>     * Error handling within `snownet` is improved. The error condition of "you gave me data on a socket I don't know about" simply no longer exists. (This reminds me I need to remove the error variant).
> 
>     * No combinatoric explosion of candidate pairs.
> 

👍 alright, yeah.

> I am happy to add the API back. Then this PR is a strict improvement over what we had before. I'll also try to model the above network topology as a docker-compose test.

I'd be okay with that.

> This might be a silly question but what is the point of using firezone in this topology? You are already in the same network as your gateway so you should be able to reach all the resources.

There are a few reasons why:
* Rarely, a gateway might have another interface that can reach the resources
* More commonly, the user always logs into firezone, since not all resources are available locally and can be using it on different networks to different gateways, some of which are within these conditions

**Comment by jamilbk** [comment]
> This might be a silly question but what is the point of using Firezone in this topology?

I think this gets to the heart of the discussion.

Firezone is only applicable if there's a stateful firewall allowing egress in between the client and gateway that need to communicate. Otherwise, what's the point?

In many cases, there's only one stateful firewall, especially for user clients. In cloud topologies, data centers, and on-prem, however, networks are commonly more complex. It's not uncommon to have topologies like this:

<p align="center">

</p>

In the topology above, the Firewall between App Servers and DB would need to be configured to explicitly allow traffic. Firezone wouldn't be useful.

**But** keep in mind that host firewalls are **also** stateful firewalls.

So with Firezone, instead of managing lots of firewalls in between each of your networks that need to communicate, you just enable the host firewall (`ufw enable`) for each host, deny all ingress, and hole-punch between each host directly. Hosts aren't allowed to talk to each other by default anymore (so called "zero trust"). Firezone becomes the way hosts talk to each other. Each network flow is authorized and logged using a central set of policies under your control.

That's the vision at least. Reality will be messier.

So with the candidate gathering change introduced in this PR, communication can only happen between two peers' nearest WAN ancestor as opposed to their nearest firewalled sibling, and in some cases not allowed at all (and thus Relayed).

To be clear -- I think the algorithm introduced in this PR is fine for now. I just wanted to (hopefully) demonstrate above the broader problem this product is solving, even if we're not aiming for these use cases just yet.

### tl;dr

All of the above is to say, I think I'm agreement:

* We should bias towards shipping `snownet` ASAP
* This is a step forward and we should move ahead with this API
* Keep the old API in place, but unused (with an appropriately detailed comment)

**Comment by thomaseizinger** [comment]
Awesome! Thank you for the detailed explanation. That makes a lot of sense! :)

Today, we only have clients and gateways and only clients can connect to gateways. Unless I am misunderstanding, then the above diagram suggests that we might eventually have more of a peer-to-peer system where each node can establish connections and also allow other nodes to connect to them. Think multiple (micro) services talking to each other.

> * Like I said in the previous comment, adding all interfaces on the gateway is enough to fix it. We can avoid a lot of (likely rubbish) candidates on the client.

I have to correct myself here, the above is wrong. We will need host candidates on either end because the local host candidate is the "source" of the traffic and it also needs to target the right interface.

The above topology is very interesting and I think we should explicitly write tests that demonstrate that we can establish a direct connection in that case. Today, that would boil down to correctly using `snownet` by manually adding that extra candidate. I am curious if we can do better here and somehow discover those reliably.

**Comment by jamilbk** [comment]
> the above diagram suggests that we might eventually have more of a peer-to-peer system where each node can establish connections and also allow other nodes to connect to them. Think multiple (micro) services talking to each other.

From a graph-theory perspective, yes. A peer-to-peer system would be a digraph where each node can connect to another and vice-versa, and imply that if node A can connect to node B, then B can connect to node A. All edges are bidirectional.

I want to be very clear of an important nuance though: in Firezone (and zero trust networks), we do not want to assume that the converse is also true. That is, by default, edges in our graph are _not_ bidirectional. In the example above, it would be suspicious if the DB node _initiated_ a connection to the App server.

To build such a digraph, we have a couple options. The first is that we can consider two different color nodes. One color node is the "from" node (client), and the other a "to" node (gateway). With these constructs we can build a digraph with all appropriate edges to allow desired communication. In reality, today, that means the possibility running a client and gateway on the same host (possible now if we use `tun-firezone{0..n}` etc and handle routing loops). We could optimize this down the road by combining the headless client and gateway into a "headless agent" that can function in client or gateway mode. But I digress, that's a lot of work, and I don't want to get distracted optimizing for that now.

The other way to build such a graph is to consider every node the same color, able to both initiate and receive communication. This bidirectional (by default) digraph is the model Tailscale uses. It then requires an ACL to explicitly state which nodes can talk to each other based on attributes the node has, and which direction the edge points.

Both approaches lead to an appropriate digraph that satisfies desired communication; our bet is an optimization that assumes most nodes will only need to be one color or another, which simplifies a lot of things from a UX perspective.

**Comment by thomaseizinger** [comment]
> our bet is an optimization that assumes most edges will only need to be one color or another, which simplifies a lot of things from a UX perspective.

Makes sense. I do like the simplification too, let's hope it pans out :crossed_fingers:

File: rust/connlib/snownet/src/node.rs [hunk 2]
```diff
@@ -119,10 +121,33 @@ where
...
+    ///      │        └──────────┘       │
+    /// ┌────┴─────┐                  ┌──┴───┐
+    /// │    FW    │                  │  FW  │
+    /// └────┬─────┘                  └──┬───┘
+    ///      │            ┌──┐           │
+    ///  ┌───┴─────┐      │  │         ┌─┴──┐
+    ///  │ BACKEND ├──────┤FW├─────────┤ DB │
+    ///  └─────────┘      │  │         └────┘
+    ///                   └──┘
+    /// ```
```
Comment (jamilbk) [inline rust/connlib/snownet/src/node.rs:146]: Nice ascii art skills

**Review by jamilbk — APPROVED** [review]

File: rust/connlib/snownet/src/node.rs (hunk 2 above)
Comment (thomaseizinger) [reply rust/connlib/snownet/src/node.rs:146]: Credit to https://asciiflow.com.
